circle.yml

```diff
@@ -17,21 +15,14 @@ dependencies:
         "$CIRCLE_BUILD_URL" > version.json
     - cp version.json $CIRCLE_ARTIFACTS
     - docker info
-    # use circleci's docker cache workaround
-    - mkdir -p ~/cache/
-    - if [ -e ~/cache/docker/image.tar ]; then echo "Loading image.tar"; docker load -i ~/cache/docker/image.tar || rm ~/cache/docker/image.tar; fi
     # build image
     - docker build -t morgoth:build .
-    # Clean up old image and save the new one
-    - >
-      mkdir -p ~/cache/docker;
-      rm -f ~/cache/docker/image.tar;
-      docker save morgoth:build > ~/cache/docker/image.tar;
-      ls -l ~/cache/docker
 
 test:
   pre:
     - chmod -R 777 $CIRCLE_TEST_REPORTS
+    # Install dev dependencies
+    - bin/ci/docker-run.sh pip install -r requirements-dev.txt --require-hashes
```

Comment:
This doesn't work because the app user (which the Dockerfile specifies) doesn't have permission to install Python packages. You could fix this by saying `--user root` in this command.

I don't think this is a good idea. For one, this creates a second docker image that isn't related to the one you built, so future test runs won't include it. You could fix that by tagging the result of this command, and running your tests in that. However, then you are running the tests on a version of the docker file you aren't shipping to prod, so there may be issues that arise in prod that the tests seemingly validated.

I would simply install all dependencies needed to run the tests in the Dockerfile. It produces slightly larger images, but the reliability is worth it. It make make sense to include tools like hashin that are only used in development and not for tests or prod in a separate file.